lumicks/pylake/kymotracker/detail/gaussian_mle.py

```diff
@@ -0,0 +1,176 @@
+    """Calculates the derivatives of the peak expectation w.r.t. each parameter.
+
+    Results are returned as 2D arrays with shape [len(x), len(parameter)]"""
+    g = gaussian(x, amplitude, center, scale)
+    x_center_diff = col(x) - row(center)
+
+    d_damplitude = g / amplitude
+    d_dcenter = g * x_center_diff / row(scale) ** 2
+
+    d_dscale_numerator = (lambda x: col(x.sum(axis=1))) if len(scale) == 1 else (lambda x: (x))
+    d_dscale = d_dscale_numerator(g * x_center_diff ** 2) / scale ** 3
+
+    d_doffset = col(np.ones(g.shape).sum(axis=1)) if len(offset) == 1 else np.ones(g.shape)
+
+    return d_damplitude, d_dcenter, d_dscale, d_doffset
+
+
+def _extract_params(params, n_frames, shared_variance, shared_offset):
+    """Separates single array of parameters into individual components."""
+    n_var = 1 if shared_variance else n_frames
+    n_off = 1 if shared_offset else n_frames
+    assert len(params) == 2 * n_frames + n_var + n_off
+
+    amplitude = params[:n_frames]
+    center = params[n_frames : 2 * n_frames]
+
+    scale_slice = 2 * n_frames if shared_variance else slice(2 * n_frames, 3 * n_frames)
+    offset_start = 2 * n_frames + 1 if shared_variance else 3 * n_frames
+
+    scale = np.atleast_1d(params[scale_slice])
+    offset = params[offset_start:]
+
+    return amplitude, center, scale, offset
+
+
+def poisson_log_likelihood(params, x, photon_count, shared_variance=False, shared_offset=False):
+    """Calculates the log likelihood of Poisson distributed values.
+
+    Parameters
+    __________
+    params : array-like
+        Model parameters to be estimated; order is [amplitude, center, scale, offset]
+    x : array-like
+        Data at which the function is to be evaluated.
+    photon_count : array-like
+        Signal data to be fitted with shape [len(x), # of frames]
+    shared_variance : bool
+        If the variance is a single value shared across all frames
+    shared_offset : bool
+        If the offset is a single value shared across all frames
+    """
+    n_frames = photon_count.shape[1]
+    amplitude, center, scale, offset = _extract_params(
+        params, n_frames, shared_variance, shared_offset
+    )
+
+    expectation = peak_expectation(x, amplitude, center, scale, offset)
+    log_likelihood = photon_count * np.log(expectation) - expectation
+    return np.sum(-log_likelihood)
+
+
+def poisson_log_likelihood_jacobian(
+    params, x, photon_count, shared_variance=False, shared_offset=False
+):
+    """Calculates the derivatives of the log likelihood w.r.t. each parameter."""
+    n_frames = photon_count.shape[1]
+    amplitude, center, scale, offset = _extract_params(
+        params, n_frames, shared_variance, shared_offset
+    )
+
+    tile = lambda x: np.hstack(
+        (
+            x,
+            x,
+            col(x.sum(axis=1)) if len(scale) == 1 else x,
+            col(x.sum(axis=1)) if len(offset) == 1 else x,
```

Comment:
It took me a while to understand what's going on here. Some additional comments would be helpful.

Correct me if I'm wrong, but I'm not convinced that summing here is correct. You sum the expectations and photon counts for the aggregated sigmas and offsets prematurely I think. The likelihood is not linear, so you shouldn't sum results for the different frames before evaluating the likelihood for each frame.

If you test with a Jacobian with different model parameters per track you easily get relative errors in your Jacobian entries of more than 10%.

Reply:
I also think it would be nice to extract the functions that map parameter derivatives out of the objective function, so that they can be tested in isolation. Things become hard to test if you need all the pieces in order to test it.

But maybe that's something for a future PR.

Reply:
See https://github.com/lumicks/pylake/tree/gau_loc_not_same for an example of what I mean. First commit is the test update (to use different parameters); second is the order of operation "fix". Note that this was just a quick and dirty hack, and that I'm not proposing this as an implementation of course 😅